### PR TITLE
Fixed #25426 -- RuntimeError pickling SimpleLazyObject

### PR DESCRIPTION
**Pull request by knifenomad**
I worked on this patch because nobody seems interested in this issue. The patch is for saving django version info as unpickler's argument instead of setting it as model instance's attribute which results in RuntimeError. This happens on django 1.8.x with both python3.x/2.x.
please see https://code.djangoproject.com/ticket/25426 for details

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
An alternate implementation seems to be: https://github.com/django/django/pull/5276. Would you mind discussing the best implementation with that author?

**Comment by benjaminjkraft** [comment]
@timgraham Thanks for the ping.  I think the right thing is actually do both.

@knifenomad's patch seems like a great idea to me -- setting an attribute on the model while pickling it doesn't seem very robust, and I think this is definitely a more supported way of doing things.  I think with some `__dict__` manipulation one could come up with a case involving a `Model` but not a `SimpleLazyObject` that breaks in the current code and is not fixed by #5276, something like the following:
```
model._self = model;
model_dict = model.__dict__
model_pickle.dumps(model_dict)
```
(While pickling the dict, it will come across the model, and start pickling that, causing it to add an attribute to the dict, causing the error.)  Arguably not an operation that it matters to support, but we might as well do the more robust thing.

But I don't think this patch actually fixes #25389; it will fix the error, but I think it still causes the version check to be skipped, because `SimpleLazyObject` just doesn't call custom `__reduce__` methods.  So the test case I wrote in #5276 will still fail, and I believe pickling and unpickling a model wrapped in a `SimpleLazyObject` will still warn about the version check.  And while modifying the `__dict__` in the `__reduce__` method seems questionable to me, I don't think it's prohibited, so somebody else could define a class that does it, and `SimpleLazyObject` should ideally support that, which #5276 would do.

File: tests/utils_tests/test_simplelazyobject.py [hunk 1]
```diff
@@ -29,3 +35,16 @@ def test_pickle_py2_regression(self):
...
             cPickle.dumps(x)
+
+    def test_pickle_model_instance_with_one_to_one_field(self):
+
+        # See ticket #25426
+        user = User.objects.create_user('johndoe', 'john@example.com', 'pass')
+        UserProfile.objects.create(user=user)
+
+        x = SimpleLazyObject(lambda: user)
+        x.profile  # accsss related object
```
Comment (benjaminjkraft) [inline tests/utils_tests/test_simplelazyobject.py]: It wouldn't hurt to do two tests here, one where you access the model fields before pickling it, and one where you don't.

File: tests/utils_tests/test_simplelazyobject.py [hunk 2]
```diff
@@ -29,3 +35,16 @@ def test_pickle_py2_regression(self):
...
+
+        # This would fail with "RuntimeError: dictionary changed size during iteration",
+        # on django 1.8 both on python3.x and 2.x
+        pickle.dumps(x)  # Fails
```
Comment (benjaminjkraft) [inline tests/utils_tests/test_simplelazyobject.py]: This should also unpickle the model, and make sure there is no warning.  Or I can add that in #5276.

**Comment by knifenomad** [comment]
Great. I agree with @benjaminjkraft. Calling _wrapped.__reduce_ex__() seems much nicer way to prevent pickling issue in general way and also for detecting version warnings. for the __dict__ part, I would keep this patch for the same reason @benjaminjkraft explained. @benjaminjkraft please update the test as you like. thanks I've learned a lot.

**Comment by knifenomad** [comment]
I was nearly asleep when I posted last time. I've just reviewed @benjaminjkraft's fix again in the morning and I'm sure it's better way to go. I was worried about touching __dict__ but this becomes another issue when @benjaminjkraft's fix is shipped. I'd rather create another ticket where this fix becomes necessary :)

**Pull request closed without merging**